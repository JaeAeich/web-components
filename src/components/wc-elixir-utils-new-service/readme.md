# wc-elixir-utils-new-service



<!-- Auto Generated Below -->


## Properties

| Property    | Attribute    | Description | Type     | Default     |
| ----------- | ------------ | ----------- | -------- | ----------- |
| `apiUrl`    | `api-url`    |             | `string` | `undefined` |
| `authToken` | `auth-token` |             | `string` | `undefined` |
| `schema`    | `schema`     |             | `string` | `undefined` |


## Dependencies

### Used by

 - [app-new-service](../app-new-service)
 - [wc-elixir-service](../wc-elixir-service)

### Graph
```mermaid
graph TD;
  app-new-service --> wc-elixir-utils-new-service
  wc-elixir-service --> wc-elixir-utils-new-service
  style wc-elixir-utils-new-service fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
