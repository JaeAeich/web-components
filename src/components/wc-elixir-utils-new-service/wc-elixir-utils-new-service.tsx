import { Component, Host, h, Prop, State } from "@stencil/core";
import axios from "axios";

@Component({
  tag: "wc-elixir-utils-new-service",
  styleUrl: "wc-elixir-utils-new-service.css",
  scoped: true,
})
export class WcElixirUtilsNewService {
  @Prop() apiUrl: string;
  @Prop() schema: string;
  @Prop() authToken: string;
  @State() fields: any;
  @State() data: string = "loading";

  componentWillLoad = async () => {
    try {
      const res = await axios.get(this.apiUrl);
      this.fields = res.data.components.schemas[this.schema];
      this.data = "loaded";
    } catch (e) {
      this.data = "error";
    }
  };

  toTitleCase = (str) => {
    return str.replace(/\w\S*/g, function (txt) {
      return txt.charAt(0).toUpperCase() + txt.substr(1);
    });
  };

  isRequired = (field, fields) => {
    let index = fields.required.findIndex((req) => field === req);
    if (index == -1) {
      return false;
    }
    return true;
  };

  renderFields = () => {
    if (this.data == "loading") {
      return (
        <div class="text-center">
          <div class="text-gray-700">Loading...</div>
        </div>
      );
    }
    if (this.data == "error" || !this.fields) {
      return (
        <div class="flex text-red-400 justify-center font-semibold">
          <svg
            xmlns="http://www.w3.org/2000/svg"
            class="h-6 w-6 mr-2"
            fill="none"
            viewBox="0 0 24 24"
            stroke="currentColor"
          >
            <path
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              d="M12 8v4m0 4h.01M21 12a9 9 0 11-18 0 9 9 0 0118 0z"
            />
          </svg>
          Sorry, the open api endpoint doesn't exist.
        </div>
      );
    }
    return (
      <div>
        {Object.keys(this.fields.properties).map((property) => {
          return (
            <div class="py-2 border-b ">
              <div class="flex justify-between items-center">
                <div class="text flex-1">
                  {this.toTitleCase(property)}
                  {this.isRequired(property, this.fields) ? "*" : ""} :
                </div>
                <div class="flex-1">
                  <input
                    required={this.isRequired(property, this.fields)}
                    // placeholder={this.fields.properties[property].description}
                    class={`w-full text-sm border-2 py-1.5 px-3 focus:outline-none rounded-lg focus:shadow ${
                      this.fields.properties[property].properties ? "invisible" : ""
                    }`}
                  ></input>
                </div>
              </div>
              {this.fields.properties[property].properties ? (
                Object.keys(this.fields.properties[property].properties).map((subproperty) => {
                  return (
                    <div class="flex justify-between items-center my-2">
                      <div class="text flex-1">{this.toTitleCase(subproperty)}:</div>
                      <div class="flex-1">
                        <input
                          required={this.isRequired(property, this.fields.properties[property])}
                          class="w-full text-sm border-2 py-1.5 px-3 focus:outline-none rounded-lg focus:shadow"
                        ></input>
                      </div>
                    </div>
                  );
                })
              ) : (
                <div></div>
              )}
            </div>
          );
        })}
        <div class="text-center">
          <button
            type="submit"
            class="bg-secondary rounded-lg px-4 py-2 md:mr-2 my-2 text-white hover:shadow-lg focus:outline-none w-full md:w-48"
          >
            Submit
          </button>
        </div>
      </div>
    );
  };

  render() {
    console.log(this.fields);
    return (
      <Host>
        <form>{this.renderFields()}</form>
      </Host>
    );
  }
}
