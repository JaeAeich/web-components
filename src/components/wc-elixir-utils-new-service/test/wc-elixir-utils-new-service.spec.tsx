import { newSpecPage } from '@stencil/core/testing';
import { WcElixirUtilsNewService } from '../wc-elixir-utils-new-service';

describe('wc-elixir-utils-manage-permissions', () => {
  it('does builds', () => {
    expect(new WcElixirUtilsNewService()).toBeTruthy();
  });

  describe('has props', () => {
    it('has authToken prop', async () => {
      const page = await newSpecPage({
        components: [WcElixirUtilsNewService],
        html: '<div></div>',
      });

      let component = page.doc.createElement('wc-elixir-utils-new-service');
      component.authToken = 'component-demo';
      page.root.appendChild(component);
      await page.waitForChanges();

      expect(page.rootInstance.authToken).toBe('component-demo');
    });

    // it('has itemsPerPage prop', async () => {
    //   const page = await newSpecPage({
    //     components: [WcElixirUtilsNewService],
    //     html: '<div></div>',
    //   });

    //   let component = page.doc.createElement('wc-elixir-utils-new-service');
    //   component.authToken = 'component-demo';
    //   component.itemsPerPage = 10;
    //   page.root.appendChild(component);
    //   await page.waitForChanges();

    //   expect(page.rootInstance.itemsPerPage).toBe(10);
    // });
  });

  it('renders', async () => {
    const page = await newSpecPage({
      components: [WcElixirUtilsNewService],
      html: '<div></div>',
    });

    let component = page.doc.createElement('wc-elixir-utils-new-service');
    component.authToken = 'component-demo';
    page.root.appendChild(component);
    await page.waitForChanges();

    expect(page).toMatchSnapshot();
  });
});
