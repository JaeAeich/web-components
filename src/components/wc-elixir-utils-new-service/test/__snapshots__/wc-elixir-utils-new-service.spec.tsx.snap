// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`wc-elixir-utils-manage-permissions renders 1`] = `
Object {
  "body": <body>
  <div>
    <wc-elixir-utils-new-service>
      <form>
        <div class="flex font-semibold justify-center text-red-400">
          <svg class="h-6 mr-2 w-6" fill="none" stroke="currentColor" viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg">
            <path d="M12 8v4m0 4h.01M21 12a9 9 0 11-18 0 9 9 0 0118 0z" stroke-linecap="round" stroke-linejoin="round" stroke-width="2"></path>
          </svg>
          Sorry, the open api endpoint doesn't exist.
        </div>
      </form>
    </wc-elixir-utils-new-service>
  </div>
</body>,
  "build": Object {
    "allRenderFn": true,
    "appendChildSlotFix": false,
    "asyncLoading": true,
    "asyncQueue": true,
    "attachStyles": false,
    "cloneNodeFix": false,
    "cmpDidLoad": true,
    "cmpDidRender": true,
    "cmpDidUnload": true,
    "cmpDidUpdate": true,
    "cmpShouldUpdate": true,
    "cmpWillLoad": true,
    "cmpWillRender": true,
    "cmpWillUpdate": true,
    "connectedCallback": true,
    "constructableCSS": true,
    "cssAnnotations": false,
    "cssVarShim": true,
    "devTools": false,
    "disconnectedCallback": true,
    "dynamicImportShim": false,
    "element": true,
    "event": true,
    "hasRenderFn": true,
    "hostListener": true,
    "hostListenerTarget": true,
    "hostListenerTargetBody": true,
    "hostListenerTargetDocument": true,
    "hostListenerTargetParent": true,
    "hostListenerTargetWindow": true,
    "hotModuleReplacement": false,
    "hydrateClientSide": false,
    "hydrateServerSide": false,
    "hydratedAttribute": false,
    "hydratedClass": true,
    "initializeNextTick": true,
    "isDebug": false,
    "isDev": true,
    "isTesting": true,
    "lazyLoad": true,
    "lifecycle": true,
    "lifecycleDOMEvents": true,
    "member": true,
    "method": true,
    "mode": true,
    "observeAttribute": true,
    "profile": true,
    "prop": true,
    "propBoolean": true,
    "propMutable": true,
    "propNumber": true,
    "propString": true,
    "reflect": true,
    "safari10": false,
    "scoped": true,
    "scriptDataOpts": false,
    "shadowDelegatesFocus": true,
    "shadowDom": true,
    "shadowDomShim": false,
    "slot": true,
    "slotChildNodesFix": false,
    "slotRelocation": true,
    "state": true,
    "style": true,
    "svg": true,
    "taskQueue": true,
    "transformTagName": true,
    "updatable": true,
    "vdomAttribute": true,
    "vdomClass": true,
    "vdomFunctional": true,
    "vdomKey": true,
    "vdomListener": true,
    "vdomPropOrAttr": true,
    "vdomRef": true,
    "vdomRender": true,
    "vdomStyle": true,
    "vdomText": true,
    "vdomXlink": true,
    "watchCallback": true,
  },
  "doc": ,
  "flushLoadModule": [Function],
  "flushQueue": [Function],
  "setContent": [Function],
  "styles": Map {},
  "waitForChanges": [Function],
  "win": [MockWindow],
}
`;
